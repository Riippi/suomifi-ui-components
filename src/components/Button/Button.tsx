import React, {
  Component,
  MouseEvent,
  KeyboardEvent,
  ReactNode,
  Fragment,
} from 'react';
import { HtmlButton, HtmlButtonProps } from '../../reset';
import { VisuallyHidden } from '../Visually-hidden/Visually-hidden';
import classnames from 'classnames';
import styled from 'styled-components';
import { disabledCursor } from '../utils/css';

const baseClassName = 'fi-button';
const disabledClassName = `${baseClassName}--disabled`;

export interface MouseNonFocusProps {
  handleClick: (
    e: React.MouseEvent<Element> | React.KeyboardEvent<Element>,
  ) => void;
}

export interface MouseNonFocusReturnProps {
  onMouseDown: ((event: React.MouseEvent<HTMLButtonElement>) => void);
  onKeyUp: ((event: React.KeyboardEvent<HTMLButtonElement>) => void);
}

/**
 * @typedef {Object} buttonClick
 * @property {function} onMouseDown given button property
 * @property {function} onKeyUp given button property
 */
/** Prevent button :focus on mouse use, allow only keyboard
 * @param {function} props.handleClick function to run on click
 * @returns {buttonClick}
 */
export const mouseNonFocus = ({
  handleClick,
}: MouseNonFocusProps): MouseNonFocusReturnProps => {
  const doClick = (
    e: MouseEvent<HTMLButtonElement> | KeyboardEvent<HTMLButtonElement>,
  ) => {
    if (!!e) {
      e.preventDefault();
    }
    handleClick(e);
  };
  const keyboard = { enter: 13, space: 32 };
  const ifKeyboard = (e: KeyboardEvent<HTMLButtonElement>) =>
    (e.keyCode === keyboard.enter || e.keyCode === keyboard.space) &&
    doClick(e);

  return { onMouseDown: doClick, onKeyUp: ifKeyboard };
};
export interface ButtonProps extends AssertiveProps, HtmlButtonProps {
  /** Custom classname to extend or customize */
  className?: string;
  /** Event handler to execute when clicked
   *  @default void
   */
  onClick?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * Button element content
   */
  children?: ReactNode;
  /**
   * Define a label if children content does not indicate the button purpose,
   * alternatively you can define aria-labelledby with label-element id
   */
  'aria-label'?: string;
  /**
   * Prevent :focus when used with mouse
   */
  mouseNoFocus?: boolean;
}

interface AssertiveProps {
  /** Disable Button usage */
  disabled?: boolean;
  /** aria-live text for why Button is disabled */
  ariaAssertiveDisabled?: string;
  /** aria-live text when Button changed from disabled to enabled */
  ariaAssertiveUnDisabled?: string;
}

export class Assertive extends Component<AssertiveProps> {
  state = { unDisabled: false };

  componentDidUpdate(prevProps: AssertiveProps) {
    const { disabled } = this.props;
    if (disabled !== prevProps.disabled) {
      this.setState({ unDisabled: !disabled });
    }
  }

  render() {
    const { ariaAssertiveDisabled, ariaAssertiveUnDisabled } = this.props;
    const { unDisabled } = this.state;
    if (!ariaAssertiveDisabled && !ariaAssertiveUnDisabled) return null;
    const text = !!unDisabled ? ariaAssertiveUnDisabled : ariaAssertiveDisabled;
    return !!text ? (
      <VisuallyHidden aria-live="assertive">{text}</VisuallyHidden>
    ) : null;
  }
}

const StyledHtmlButton = styled((props: ButtonProps) => (
  <HtmlButton {...props} />
))`
  &.${disabledClassName} {
    ${disabledCursor}
  }
`;

export class Button extends Component<ButtonProps> {
  render() {
    const {
      className,
      disabled = false,
      mouseNoFocus,
      onClick,
      ariaAssertiveDisabled,
      ariaAssertiveUnDisabled,
      ...passProps
    } = this.props;
    const ifMouseNoFocus =
      !!mouseNoFocus && !!onClick
        ? () => {
            const doClick = mouseNonFocus({ handleClick: onClick });
            return {
              onMouseDown: doClick.onMouseDown,
              onKeyUp: doClick.onKeyUp,
            };
          }
        : () => ({ onClick });
    const onClickProps = !!disabled ? {} : ifMouseNoFocus();
    return (
      <Fragment>
        <StyledHtmlButton
          {...passProps}
          {...onClickProps}
          aria-disabled={disabled}
          tabIndex={0}
          className={classnames(baseClassName, className, {
            [disabledClassName]: !!disabled,
          })}
        />
        <Assertive
          disabled={disabled}
          ariaAssertiveDisabled={ariaAssertiveDisabled}
          ariaAssertiveUnDisabled={ariaAssertiveUnDisabled}
        />
      </Fragment>
    );
  }
}
