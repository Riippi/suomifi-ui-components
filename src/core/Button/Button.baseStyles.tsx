import { css } from '@emotion/core';
import { suomifiTheme, ThemeProp } from '../theme';
import { ButtonProps } from './Button';
import { element, focus, button } from '../theme/reset';

const fullWidthStyles = css`
  display: block;
  width: 100%;
`;

const negativeStyles = (theme: ThemeProp) => css`
  &.fi-button--negative {
    background: none;
    background-color: ${theme.colors.highlightBase};
    border: 1px solid ${theme.colors.whiteBase};
    text-shadow: none;

    &:hover {
      background: ${theme.gradients.whiteBaseNegative};
    }

    &:active {
      background: none;
      background-color: ${theme.colors.highlightBase};
    }

    &.fi-button--disabled,
    &[disabled],
    &:disabled {
      opacity: 0.5;
      background: none;
    }
  }
`;

const secondary = (theme: ThemeProp) => css`
  color: ${theme.colors.highlightBase};
  background: none;
  background-color: ${theme.colors.whiteBase};
  border: 1px solid ${theme.colors.highlightBase};
  text-shadow: none;

  &:hover {
    background: ${theme.gradients.depthLight26};
  }

  &:active {
    background: none;
    background-color: ${theme.colors.highlightLight53};
  }

  &.fi-button--disabled,
  &[disabled],
  &:disabled {
    color: ${theme.colors.depthBase};
    text-shadow: none;
    background: none;
    background-color: ${theme.colors.highlightLight53};
    border-color: ${theme.colors.depthBase};
  }
`;

const secondaryStyles = (theme: ThemeProp) => css`
  &.fi-button--secondary {
    ${secondary(theme)}
  }
`;

const secondaryNoBorderStyles = (theme: ThemeProp) => css`
  &.fi-button--secondary-noborder {
    ${secondary(theme)}
    border: none;
  }
`;

const tertiaryStyles = (theme: ThemeProp) => css`
  &.fi-button--tertiary {
    ${secondary(theme)}
    background: ${theme.colors.highlightLight50};
    border: none;

    &:hover {
      background: ${theme.colors.highlightLight53};
    }

    &:active {
      background: ${theme.colors.highlightLight53};
    }
  }
`;

export const baseStyles = ({
  theme = suomifiTheme,
  fullWidth = false,
}: ButtonProps) => css`
  ${button(theme)}
  padding: ${theme.spacing.s} ${theme.spacing.m};
  min-height: 40px;
  color: ${theme.colors.whiteBase};
  background: ${theme.gradients.highlightBase};
  border-radius: ${theme.radius.basic};
  text-align: center;
  text-shadow: ${theme.shadows.invertTextShadow};
  cursor: pointer;

  &:hover {
    background: ${theme.gradients.highlightLight4};
  }

  &:active {
    background: ${theme.gradients.highlightDark9};
  }

  &.fi-button--disabled,
  &[disabled],
  &:disabled {
    text-shadow: 0 1px 1px ${theme.colors.blackBase};
    background: ${theme.gradients.depthBase};
    pointer-events: none;
    user-select: none;
  }

  ${fullWidth && fullWidthStyles}
  ${negativeStyles(theme)}
  ${secondaryStyles(theme)}
  ${secondaryNoBorderStyles(theme)}
  ${tertiaryStyles(theme)}

  & > .fi-button_icon {
    width: 16px;
    height: 16px;
    margin-right: ${theme.spacing.s};
    vertical-align: middle;
    transform: translateY(-0.1em);
    &.fi-button_icon--right {
      margin-right: 0;
      margin-left: ${theme.spacing.s};
    }
  }
`;

export const unStyled = ({ theme = suomifiTheme }) => css`
  ${element(theme)}
  ${focus(theme)}
  border-radius: ${theme.radius.basic};
  cursor: pointer;
`;
