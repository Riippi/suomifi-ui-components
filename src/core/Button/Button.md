```js
import { Button } from 'suomifi-ui-components';

<>
  <Button
    className="my-button--test"
    onClick={() => console.log('Test button click')}
    type="submit"
  >
    Button
  </Button>

  <Button disabled fullWidth>
    Button disabled fullWidth
  </Button>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <Button icon="login" aria-label="Login">
    Button icon="login" aria-label="Login"
  </Button>

  <Button iconRight="login" aria-labelledby="button-label">
    <span id="button-label">Login</span> iconRight="login"
    aria-labelledby="button-label"
  </Button>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <div example="negative">
    <Button.negative>Button.negative</Button.negative>

    <Button.negative disabled fullWidth icon="login">
      Button.negative disabled fullWidth icon="login"
    </Button.negative>
  </div>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <Button.secondary>Button.secondary</Button.secondary>

  <Button.secondary icon="login">
    Button.secondary icon="login"
  </Button.secondary>

  <Button.secondary disabled fullWidth icon="login">
    Button.secondary disabled fullWidth icon="login"
  </Button.secondary>

  <Button.secondaryNoborder>
    Button.secondaryNoborder
  </Button.secondaryNoborder>

  <Button.secondaryNoborder icon="login">
    Button icon="login"
  </Button.secondaryNoborder>

  <Button.secondaryNoborder disabled icon="login">
    Button.secondaryNoborder disabled icon="login"
  </Button.secondaryNoborder>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <Button.tertiary>Button.tertiary</Button.tertiary>

  <Button.tertiary disabled icon="login">
    Button.tertiary disabled icon="login"
  </Button.tertiary>
</>;
```
