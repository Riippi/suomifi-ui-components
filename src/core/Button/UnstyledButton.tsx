import React from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { unStyled } from './Button.baseStyles';
import { InternalTokensProp } from '../theme';
import { Button, ButtonProps } from '../../components/Button/Button';

const baseClassName = 'fi-button';

export const UnstyledButton = styled(
  (props: ButtonProps & InternalTokensProp) => {
    const { className, ...passProps } = props;
    return (
      <Button {...passProps} className={classnames(className, baseClassName)} />
    );
  },
)`
  ${props => unStyled(props)}
`;
