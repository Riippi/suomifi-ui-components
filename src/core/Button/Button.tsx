import React, { Component } from 'react';
import styled from '@emotion/styled';
import classnames from 'classnames';
import { ThemeComponent, ThemeProp } from '../theme';
import { withDefaultTheme } from '../theme/utils';
import { baseStyles } from './Button.baseStyles';
import {
  Button as CompButton,
  ButtonProps as CompButtonProps,
} from '../../components/Button/Button';
import { Icon, IconProps, IconKeys } from '../Icon/Icon';
import { UnstyledButton } from './UnstyledButton';

type ButtonVariant =
  | 'default'
  | 'negative'
  | 'secondary'
  | 'secondary-noborder'
  | 'tertiary'
  | 'unstyled';

export interface ButtonProps extends CompButtonProps, ThemeComponent {
  /**
   * Set width to grow all available space
   */
  fullWidth?: boolean;
  /**
   * 'default' | 'negative' | 'secondary' | 'secondary-noborder' | 'tertiary'
   * @default default
   */
  variant?: ButtonVariant;
  /**
   * Icon from suomifi-theme
   */
  icon?: IconKeys;
  /**
   * Icon from suomifi-theme to be placed on right side
   */
  iconRight?: IconKeys;
  /**
   * Properties given to Icon-component
   */
  iconProps?: IconProps;
}

const baseClassName = 'fi-button';
const iconClassName = `${baseClassName}_icon`;
const iconRightClassName = `${baseClassName}_icon--right`;

const StyledButton = styled(
  ({
    theme,
    fullWidth,
    variant = 'default',
    className,
    ...passProps
  }: ButtonProps & { right?: boolean }) => (
    <CompButton
      {...passProps}
      className={classnames(className, {
        [`${baseClassName}--${variant}`]: variant !== 'default',
      })}
    />
  ),
)`
  ${props => baseStyles(props)}
`;

const iconColor = ({
  theme,
  invert,
  disabled,
}: {
  theme?: ThemeProp;
  invert?: boolean;
  disabled?: boolean;
}) => {
  if (!!theme) {
    const defaultColor = !!disabled
      ? theme.colors.depthBase
      : theme.colors.whiteBase;
    const secondaryColor = !!disabled
      ? theme.colors.depthBase
      : theme.colors.highlightBase;
    return !!invert ? secondaryColor : defaultColor;
  }
  return undefined;
};

class ButtonWithIcon extends Component<ButtonProps> {
  render() {
    const { icon, iconRight, iconProps = {}, ...passProps } = this.props;
    const { theme, disabled, variant } = passProps;
    const { className: iconPropsClassName, ...passIconProps } = iconProps;

    if (variant === 'unstyled') {
      return <UnstyledButton {...passProps} />;
    }

    const secondaryOrTertiary =
      !!variant &&
      (variant === 'secondary' ||
        variant === 'secondary-noborder' ||
        variant === 'tertiary');
    return (
      <StyledButton {...passProps}>
        {!!icon && (
          <Icon
            mousePointer={true}
            icon={icon}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            className={classnames(iconClassName, iconPropsClassName)}
            {...passIconProps}
          />
        )}
        {passProps.children}
        {!!iconRight && (
          <Icon
            mousePointer={true}
            icon={iconRight}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            className={classnames(
              iconClassName,
              iconRightClassName,
              iconPropsClassName,
            )}
            {...passIconProps}
          />
        )}
      </StyledButton>
    );
  }
}

/**
 * <i class="semantics" />
 * Use for inside Application onClick events.<br />
 * When using Button.secondaryNoborder with other than white background,<br />
 * define styles background color for all needed states (:hover, :active, :disabled)<br /><br />
 * You can use separate UnstyledButton to get Button without variant styles added by<br />
 * `import { UnstyledButton } from 'suomifi-ui-components';` or `<Button.unstyled />`.
 */
export class Button extends Component<ButtonProps> {
  static negative = (props: ButtonProps) => {
    return <ButtonWithIcon {...withDefaultTheme(props)} variant="negative" />;
  };

  static secondary = (props: ButtonProps) => {
    return <ButtonWithIcon {...withDefaultTheme(props)} variant="secondary" />;
  };

  static secondaryNoborder = (props: ButtonProps) => {
    return (
      <ButtonWithIcon
        {...withDefaultTheme(props)}
        variant="secondary-noborder"
      />
    );
  };

  static tertiary = (props: ButtonProps) => {
    return <ButtonWithIcon {...withDefaultTheme(props)} variant="tertiary" />;
  };

  static unstyled = (props: ButtonProps) => {
    return <UnstyledButton {...withDefaultTheme(props)} />;
  };

  render() {
    return <ButtonWithIcon {...withDefaultTheme(this.props)} />;
  }
}
