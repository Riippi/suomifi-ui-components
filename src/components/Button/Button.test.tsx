import React from 'react';
import { render } from '@testing-library/react';

import { Button } from './Button';

test('calling render with the same component on the same container does not remount', () => {
  const buttonRendered = render(<Button data-testid="button">Test</Button>);
  const { getByTestId, container, rerender } = buttonRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('button').textContent).toBe('Test');

  // re-render the same component with different props
  rerender(<Button data-testid="nottub">Test two</Button>);
  expect(getByTestId('nottub').textContent).toBe('Test two');
});
