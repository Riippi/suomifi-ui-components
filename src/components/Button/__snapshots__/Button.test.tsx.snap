// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.emotion-0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.emotion-0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.emotion-0::-webkit-inner-spin-button {
  height: auto;
}

.emotion-0::-webkit-outer-spin-button {
  height: auto;
}

<button
  aria-disabled="false"
  class="fi-button emotion-0"
  data-testid="button"
  tabindex="0"
  type="button"
>
  Test
</button>
`;
